nx/source/kernel/rwlock.c

```diff
@@ -1,59 +1,170 @@
 // Copyright 2018 plutoo
 #include "kernel/mutex.h"
 #include "kernel/rwlock.h"
+#include "../internal.h"
+
+NX_INLINE u32 _GetCurrentThreadTag(void) {
+    return getThreadVars()->handle;
+}
 
 void rwlockInit(RwLock* r) {
     mutexInit(&r->mutex);
-    condvarInit(&r->condvar_readers);
-    condvarInit(&r->condvar_writer);
-
-    r->readers = 0;
-    r->writer = false;
+    condvarInit(&r->condvar_reader_wait);
+    condvarInit(&r->condvar_writer_wait);
+
+    r->read_lock_count = 0;
+    r->read_waiter_count = 0;
+    r->write_lock_count = 0;
+    r->write_waiter_count = 0;
+    r->write_owner_tag = 0;
 }
 
 void rwlockReadLock(RwLock* r) {
+    const u32 cur_tag = _GetCurrentThreadTag();
+
+    if (r->write_owner_tag == cur_tag) {
+        r->read_lock_count++;
+        return;
+    }
+
     mutexLock(&r->mutex);
 
-    while (r->writer) {
-        condvarWait(&r->condvar_writer, &r->mutex);
+    while (r->write_waiter_count > 0) {
+        r->read_waiter_count++;
+        condvarWait(&r->condvar_reader_wait, &r->mutex);
+        r->read_waiter_count--;
     }
 
-    r->readers++;
+    r->read_lock_count++;
 
     mutexUnlock(&r->mutex);
 }
 
-void rwlockReadUnlock(RwLock* r) {
-    mutexLock(&r->mutex);
+bool rwlockTryReadLock(RwLock* r) {
+    const u32 cur_tag = _GetCurrentThreadTag();
 
-    if (--r->readers == 0) {
-        condvarWakeAll(&r->condvar_readers);
+    if (r->write_owner_tag == cur_tag) {
+        r->read_lock_count++;
+        return true;
+    }
+
+    if (!mutexTryLock(&r->mutex)) {
+        return false;
+    }
+
+    const bool got_lock = r->write_waiter_count == 0;
+    if (got_lock) {
+        r->read_waiter_count++;
```

Comment:
This accomplishes nothing and there's nothing decrementing the waiter count, since this thread isn't put to sleep. Probably this if section should be removed.